File: app/src/main/java/com/example/android/loginapp/viewModel/viewModelImpl/LoginViewModelImpl.java
package com.example.android.loginapp.viewModel.viewModelImpl;

import com.example.android.loginapp.dataLayer.LoginApi;
import com.example.android.loginapp.model.LoginResponse;
import com.example.android.loginapp.viewModel.LoginViewModel;

import java.util.HashMap;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class LoginViewModelImpl implements LoginViewModel {

    private LoginApi api;

    public LoginViewModelImpl(LoginApi api) {
        this.api = api;
    }

    @Override
    public Observable<LoginResponse> login(HashMap<String, String> body) {
        return api.login(body)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    /**
     * Checks if one of user inputs is empty
     * @param email input
     * @param password input
     * @return
     */
    @Override
    public Boolean checkEmptyInputs(String email, String password){
        if(email.isEmpty() || password.isEmpty()){
            return false;
        }
        return true;
    }

    /**
     * checks if email is in valid format
     * Got from "https://codereview.stackexchange.com/questions/33546/simple-code-to-check-format-of-user-inputted-email-address"
     * @param email
     * @return
     */
    @Override
    public Boolean isValidEmail(String email) {
        return email.matches("[a-zA-Z0-9\\.]+@[a-zA-Z0-9\\-\\_\\.]+\\.[a-zA-Z0-9]{3}");
    }

    /**
     * checks is password is valid(6 characters or more)
     * @param password
     * @return
     */
    @Override
    public Boolean isValidPassword(String password){
        Boolean isValid = (password.length() >= 6)? true : false;
        return isValid;
    }

}

File: app/src/main/java/com/example/android/loginapp/viewModel/viewModelImpl/LoginViewModelImpl.kt
package com.example.android.loginapp.viewModel.viewModelImpl

import com.example.android.loginapp.dataLayer.LoginApi
import com.example.android.loginapp.model.LoginResponse
import com.example.android.loginapp.util.Constants
import com.example.android.loginapp.util.LoginThrowable
import com.example.android.loginapp.viewModel.LoginViewModel
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.util.HashMap

class LoginViewModelImpl : LoginViewModel {

    private var api: LoginApi

    constructor(api: LoginApi) {
        this.api = api
    }


    override fun login(type: String?, body: HashMap<String, String>?): Observable<LoginResponse> {
        return api.login(type, body).flatMap { Response ->
            if (Response.code() == Constants.OK_STATUS_CODE) {
                Observable.just(Response.body())
            } else {
                Observable.error(LoginThrowable())
            }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
        }
    }

    /**
     * Checks if one of user inputs is empty
     */
    override fun checkEmptyInputs(email: String, password: String): Boolean {
        return !(email.isEmpty() || password.isEmpty())
    }

    /**
     * checks if email is in valid format
     * Got from "https://codereview.stackexchange.com/questions/33546/simple-code-to-check-format-of-user-inputted-email-address"
     */
    override fun isValidEmail(email: String): Boolean {
        return email.matches("[a-zA-Z0-9\\.]+@[a-zA-Z0-9\\-\\_\\.]+\\.[a-zA-Z0-9]{3}".toRegex())
    }

    /**
     * checks is password is valid(6 characters or more)
     */
    override fun isValidPassword(password: String): Boolean {
        return password.length >= 6
    }
}